### ISSUE TITLE
Enhancement req: Enable automatic restart of Java based services

### ISSUE THREAD

**Issue by bijugs** [agileorbit-cookbooks/java#356]
When Java version is changed using this cookbook, any Java based services/processes need to be restarted so that the new Java version takes in effect. By providing a facility to automatically restart those services managed through ``chef``, it will help users of this cookbook.

Here is an approach

- Add a default node attribute which can be used to store all the services which need to be restarted when there is a change in Java

```
default['java']['services'] = []
```
this attribute can be overridden by the user of the cookbook through a wrapper or any other means.

- Modify the resource ``file '/etc/profile.d/jdk.sh'`` in ``recipes/set_java_home.rb`` as in the following

```
 file '/etc/profile.d/jdk.sh' do
   content "export JAVA_HOME=#{node['java']['java_home']}"
   mode 00755
+  not_if "grep JAVA_HOME=#{node['java']['java_home']} /etc/profile.d/jdk.sh"
+  node['java']['services'].each do |service|
+    notifies :restart,"service[#{service}]",:delayed
+  end
 end
```
The modified code will generate the required notifications to the services listed in the array stored in the new node attribute added in the previous step.

**Comment by bijugs** [comment]
If the approach detailed above is reasonable, please review and merge PR #357 .

**Comment by erichelgeson** [comment]
Handling other services outside of the cookbook seems problematic, you can get the same behavior and a lot more flexibility (eg, reload, delay/instant/etc) if you use subscribes:

```ruby
service "hdfs" do
  subscribes :restart, 'file[/etc/profile.d/jdk.sh]', :immediately
end
```

**Comment by bijugs** [comment]
Thanks @erichelgeson for the suggestion. Few points for completeness 
- If users take this approach, it will put constraints on ``java`` cookbook with making any changes unless there is a guarantee that the ``file '/etc/profile.d/jdk.sh'…`` will be maintained as it is now.
- If the guarantee can't be provided, then users will have to make changes when there is a changes in ``java`` cookbook and keeping track of changes is not straight forward.
- Irrespective of the approach the ``file '/etc/profile.d/jdk.sh'…`` resource need to be changed to have the ``not_if`` condition since in the current code the resource gets executed in all the runs.

**Comment by erichelgeson** [comment]
Right, I only used the resource you provided as an example. There's no gaurntee that the `jdk.sh` file will change with a new jdk - if you did something like `/opt/jdk_current` - also openjdk system packages likely wont change their java home dir.

I'm all for a stable resource to notify of a jdk change which others could subscribe to - unsure what that could be just yet as we don't have access to previous runs node attributes. Probably have to have some file we can check against that is also cross platform.

The `file` resource should likely be changed to a `template` so you don't need to do a manual `not_if` check.

**Comment by erichelgeson** [comment]
Thanks for 355, I've merged it.

I disagree that notifies is the right solution, it locks you into the type of notification (restart in this pr) and when (delayed). A subscription puts it in the consumers hand on what/when/how it should be done. There may be other actions you'd like to take besides restarting a service. Also it is not platform independent.

As I stated, if we built a resource dedicated being the subscription point for jdk changes it would solve the problem and be flexible - I'll look into a way to provide that and create a PR later this week for you to review.
